JSDoc namespace type definitions. Named import/export not working.

<!-- 🚨 STOP 🚨 𝗦𝗧𝗢𝗣 🚨 𝑺𝑻𝑶𝑷 🚨

Half of all issues filed here are duplicates, answered in the FAQ, or not appropriate for the bug tracker. Even if you think you've found a *bug*, please read the FAQ first, especially the Common "Bugs" That Aren't Bugs section!

Please help us by doing the following steps before logging an issue:
  * Search: https://github.com/Microsoft/TypeScript/search?type=Issues
  * Read the FAQ: https://github.com/Microsoft/TypeScript/wiki/FAQ

Please fill in the *entire* template below.
-->

<!--
Please try to reproduce the issue with the latest published version. It may have already been fixed.
For npm: `typescript@next`
This is also the 'Nightly' version in the playground: http://www.typescriptlang.org/play/?ts=Nightly
-->
**SEE EDITS BELOW!**

**TypeScript Version:**  3.9.0-dev.20200314

<!-- Search terms you tried before logging this (so others can find this issue more easily) -->
**Search Terms:** JSDoc not exported. JSDoc type not available. "X" has no exported member "Y".

**Code**

./file-a.js (Exporting)
```
/**
 * @typedef {number} MyClass.arg
 */

/**
 * @classdesc Description.
 */
export class MyClass {
    /**
     * @constructs MyClass
     * @param {MyClass.arg} arg - Arg.
     */
    constructor(arg) {
        this.arg = arg;
    }
}
```

./file-b.js (Importing)
```
import {MyClass} from './file-a';

/**
 * @function myFunction
 * @description Description.
 * @param {MyClass.arg} arg - Arg.
 * @returns {number} Returns.
 */
function myFunction(arg) {
    return arg * arg;
}

myFunction({});
```

**Expected behavior:**
`MyClass.arg` type should be available to be used in JSDoc in `file-b.js`. Generated declaration files should export all namespace's JSDoc types.

**Actual behavior:**
`error TS2694: Namespace '"/some-path/file-a".MyClass' has no exported member 'arg'.`

**Playground Link:** <!-- A link to a TypeScript Playground "Share" link which demonstrates this behavior --> Not possible because multiple files involved.

**Related Issues:** <!-- Did you find other bugs that looked similar? -->
There are similar issues about type export/import problems, but none specifically addresses the above JSDocs problem.

As seen in the following screenshots, JSDoc type is correctly exported, imported and analysed by WebStorm, but not by TypeScript.

Please note that the above code generates correct documentation using `jsdoc` command. And it respects JSDoc specifications.

**Screenshots:**
<img width="1128" alt="Screenshot 2020-03-14 at 19 19 55" src="https://user-images.githubusercontent.com/34403992/76688242-660d5c80-662b-11ea-8403-ef79796f8e03.png">
<img width="1128" alt="Screenshot 2020-03-14 at 19 20 49" src="https://user-images.githubusercontent.com/34403992/76688246-6a397a00-662b-11ea-824a-1512825abfb0.png">

**IMPORTANT EDIT:**
Adding a default export to `file-a.js` like `default export MyClass;` adds the following statement to the output declaration file:
```
export declare namespace MyClass {
    export type arg = number;
}
```
So it seems that the problem is with named exports. However, it is still not being imported and correctly interpreted in `file-b.js`.

BUT, if I use default import in `file-b.js` like `import MyClass from './file-a';` everything works as expected!

So, definitely, bug is with named import/export! :)

I think that a fix should be considered to be fully compliant with the spec. If not, the current behaviour remains a little confusing.

**FINAL EDIT**
Using `export {MyClass};` at the end of the file instead of placing it in the class declaration works too! Now the thing becomes trivial. Considering namespaces as objects/members, exporting the class reference should export all its members (?).

--- a/src/compiler/binder.ts
+++ b/src/compiler/binder.ts
@@ -586,6 +586,9 @@ namespace ts {
         }
 
         function jsdocTreatAsExported(node: Node) {
+            if (node.parent && isModuleDeclaration(node)) {
+                node = node.parent;
+            }
             if (!isJSDocTypeAlias(node)) return false;
             // jsdoc typedef handling is a bit of a doozy, but to summarize, treat the typedef as exported if:
             // 1. It has an explicit name (since by default typedefs are always directly exported, either at the top level or in a container), or

--- a/tests/baselines/reference/jsDeclarationsImportAliasExposedWithinNamespace.js
+++ b/tests/baselines/reference/jsDeclarationsImportAliasExposedWithinNamespace.js
@@ -54,14 +54,6 @@ export {testFn, testFnTypes};
 
 
 //// [file.d.ts]
-/**
- * @namespace myTypes
- * @global
- * @type {Object<string,*>}
- */
-export const myTypes: {
-    [x: string]: any;
-};
 export namespace myTypes {
     type typeA = string | RegExp | (string | RegExp)[];
     type typeB = {
@@ -76,7 +68,18 @@ export namespace myTypes {
     };
     type typeC = Function | typeB;
 }
+/**
+ * @namespace myTypes
+ * @global
+ * @type {Object<string,*>}
+ */
+export const myTypes: {
+    [x: string]: any;
+};
 //// [file2.d.ts]
+export namespace testFnTypes {
+    type input = boolean | Function | myTypes.typeB;
+}
 /** @typedef {boolean|myTypes.typeC} testFnTypes.input */
 /**
  * @function testFn
@@ -85,6 +88,7 @@ export namespace myTypes {
  * @returns {number|null} Result.
  */
 export function testFn(input: testFnTypes.input): number | null;
+import { myTypes } from "./file.js";
 /**
  * @namespace testFnTypes
  * @global
@@ -93,7 +97,3 @@ export function testFn(input: testFnTypes.input): number | null;
 export const testFnTypes: {
     [x: string]: any;
 };
-export namespace testFnTypes {
-    type input = boolean | Function | myTypes.typeB;
-}
-import { myTypes } from "./file.js";

--- a/tests/baselines/reference/jsDeclarationsImportAliasExposedWithinNamespace.symbols
+++ b/tests/baselines/reference/jsDeclarationsImportAliasExposedWithinNamespace.symbols
@@ -21,7 +21,7 @@ const myTypes = {
 /** @typedef {myTypes.typeB|Function} myTypes.typeC */
 
 export {myTypes};
->myTypes : Symbol(myTypes, Decl(file.js, 19, 8))
+>myTypes : Symbol(myTypes, Decl(file.js, 19, 8), Decl(file.js, 9, 50), Decl(file.js, 12, 12), Decl(file.js, 17, 38))
 
 === tests/cases/conformance/jsdoc/declarations/file2.js ===
 import {myTypes} from './file.js';
@@ -63,5 +63,5 @@ function testFn(input) {
 
 export {testFn, testFnTypes};
 >testFn : Symbol(testFn, Decl(file2.js, 27, 8))
->testFnTypes : Symbol(testFnTypes, Decl(file2.js, 27, 15))
+>testFnTypes : Symbol(testFnTypes, Decl(file2.js, 27, 15), Decl(file2.js, 11, 37))
 

--- a/tests/baselines/reference/jsDeclarationsImportAliasExposedWithinNamespace.types
+++ b/tests/baselines/reference/jsDeclarationsImportAliasExposedWithinNamespace.types
@@ -26,7 +26,7 @@ export {myTypes};
 
 === tests/cases/conformance/jsdoc/declarations/file2.js ===
 import {myTypes} from './file.js';
->myTypes : { [x: string]: any; }
+>myTypes : any
 
 /**
  * @namespace testFnTypes
@@ -50,12 +50,12 @@ const testFnTypes = {
  */
 function testFn(input) {
 >testFn : (input: testFnTypes.input) => number | null
->input : boolean | Function | myTypes.typeB
+>input : import("tests/cases/conformance/jsdoc/declarations/file2").testFnTypes.input
 
     if (typeof input === 'number') {
 >typeof input === 'number' : boolean
 >typeof input : "string" | "number" | "bigint" | "boolean" | "symbol" | "undefined" | "object" | "function"
->input : boolean | Function | myTypes.typeB
+>input : import("tests/cases/conformance/jsdoc/declarations/file2").testFnTypes.input
 >'number' : "number"
 
         return 2 * input;

--- a/tests/baselines/reference/jsDeclarationsImportNamespacedType.js
+++ b/tests/baselines/reference/jsDeclarationsImportNamespacedType.js
@@ -1,0 +1,22 @@
+//// [tests/cases/conformance/jsdoc/declarations/jsDeclarationsImportNamespacedType.ts] ////
+
+//// [file.js]
+import { dummy } from './mod1'
+/** @type {import('./mod1').Dotted.Name} - should work */
+var dot2
+
+//// [mod1.js]
+/** @typedef {number} Dotted.Name */
+export var dummy = 1
+
+
+
+
+//// [mod1.d.ts]
+/** @typedef {number} Dotted.Name */
+export var dummy: number;
+export namespace Dotted {
+    type Name = number;
+}
+//// [file.d.ts]
+export {};

--- a/tests/baselines/reference/jsDeclarationsImportNamespacedType.symbols
+++ b/tests/baselines/reference/jsDeclarationsImportNamespacedType.symbols
@@ -1,0 +1,13 @@
+=== tests/cases/conformance/jsdoc/declarations/file.js ===
+import { dummy } from './mod1'
+>dummy : Symbol(dummy, Decl(file.js, 0, 8))
+
+/** @type {import('./mod1').Dotted.Name} - should work */
+var dot2
+>dot2 : Symbol(dot2, Decl(file.js, 2, 3))
+
+=== tests/cases/conformance/jsdoc/declarations/mod1.js ===
+/** @typedef {number} Dotted.Name */
+export var dummy = 1
+>dummy : Symbol(dummy, Decl(mod1.js, 1, 10))
+

--- a/tests/baselines/reference/jsDeclarationsImportNamespacedType.types
+++ b/tests/baselines/reference/jsDeclarationsImportNamespacedType.types
@@ -1,0 +1,14 @@
+=== tests/cases/conformance/jsdoc/declarations/file.js ===
+import { dummy } from './mod1'
+>dummy : number
+
+/** @type {import('./mod1').Dotted.Name} - should work */
+var dot2
+>dot2 : number
+
+=== tests/cases/conformance/jsdoc/declarations/mod1.js ===
+/** @typedef {number} Dotted.Name */
+export var dummy = 1
+>dummy : number
+>1 : 1
+

--- a/tests/cases/conformance/jsdoc/declarations/jsDeclarationsImportNamespacedType.ts
+++ b/tests/cases/conformance/jsdoc/declarations/jsDeclarationsImportNamespacedType.ts
@@ -1,0 +1,11 @@
+// @declaration: true
+// @emitDeclarationOnly: true
+// @checkJs: true
+// @filename: file.js
+import { dummy } from './mod1'
+/** @type {import('./mod1').Dotted.Name} - should work */
+var dot2
+
+// @filename: mod1.js
+/** @typedef {number} Dotted.Name */
+export var dummy = 1

